Port Eval Pipeline from csv backend to MetaDataset backend

jhaux opened the pull request:
This is a breaking change, old exported data cannot be loaded after this change and needs to be converted.

The following has changed:
1. No csv is created but paths are still stored as labels, now with `name:loader` key.
2. Labels are in a subfolder `model_outputs/labels`.
3. A meta.yaml is created, which contains information if an exception occured

jhaux committed:
Merge branch 'master' of https://github.com/pesser/edflow

jhaux committed:
Merge branch 'master' of https://github.com/pesser/edflow

jhaux committed:
:tada: Changes Eval Pipeline backend from EvalDatafolder to MetaDataset

Now we do not use csv files anymore, but numpy memory maps which is
faster and scales way better.

jhaux committed:
changelog

jhaux committed:
:bug: Fixes type

jhaux committed:
:tada: Adds new helper output on how to reuse the exported data

jhaux committed:
:green_heart: Fixes edeval tests

jhaux committed:
:green_heart: Fixes eval standalone tests

jhaux committed:
Merge branch 'master' into port_evaldatafolder

jhaux commented:
This closes #201

pesser commented on edflow/eval/pipeline.py:
```diff
@@ -119,6 +119,7 @@ def eval_op(self, inputs):
 from edflow.data.util import adjust_support
 from edflow.util import walk, retrieve, pop_keypath
 from edflow.data.dataset import DatasetMixin, CsvDataset, ProcessedDataset
```
Do we still need CsvDataset?

pesser commented on edflow/eval/pipeline.py:
```diff
@@ -346,20 +348,17 @@ def after_epoch(self, epoch):
-            self.data_frame = self.data_frame.sort_index()
-            self.data_frame.to_csv(csv_path, index=False)
+        if not hasattr(self, "exception_occured"):
+            had_exception = f"    .. note ::\n\n        No exception encountered during creation.\n\n"
```
Do we want this note in case no exception happened? I think the non-exceptional case requires no mentioning.

jhaux replied:
I would argue that this kind of documentation should be done to simply be sure everything went alright. It does not really cost anything and does not produce unnecessary log outputs. This information will be written to the `meta.yaml` of the created dataset, thus allowing anyone who wants to use this data to explicitly know if everything is created.

pesser commented on edflow/eval/pipeline.py:
```diff
@@ -368,11 +367,16 @@ def save_csv(self):
+        self.logger.info(
+            "To directly reuse the data simply use the following command:\n"
+            + "from edflow.data.believers.meta import MetaDataset\n"
+            + f'M = MetaDataset("{os.path.abspath(self.save_root)}"\n)'
```
Nice, this is super helpful!

pesser commented on edflow/eval/pipeline.py:
```diff
@@ -403,10 +407,12 @@ def at_exception(self, *args, **kwargs):
 
-class EvalDataFolder(DatasetMixin):
-    """ """
+class EvalDataFolder(MetaDataset):
```
should be kicked out.

jhaux replied:
Agreed! Will be done asap

pesser reviewed (changes requested):
Looking forward to have the MetaDataset integrated in the evaluation pipeline!

pesser commented:
Currently, the following example

    import numpy as np                                                                                        
    from edflow import TemplateIterator                                                                       
                                                                                                          
    class Iterator(TemplateIterator):                                                                         
        def __init__(self, *args, **kwargs):                                                                  
            super().__init__(*args, **kwargs)                                                                 
                                                                                                          
        def step_op(self, model, **batch):                                                                    
            key = list(batch.keys())[0]                                                                       
            def eval_op():                                                                                    
                return {"labels": {"point": np.zeros(batch[key].shape[0])}}                                   
            return dict(eval_op=eval_op)                                                                      
                                                                                                          

    def callback(root, data_in, data_out, config, **kwargs):                                                  
        print(data_out.labels)                                                                                


produces

    {'oint': memmap([0., 0., 0., ..., 0., 0., 0.])}

Somewhere, the 'p' is deleted. Please check that this PR fixes it.

jhaux committed:
:fire: Cleanup of eval pipeline code

Removes old loading functionality. We should probably think about adding
a conversion script of the old style outputs to the new ones.

jhaux commented:
> ...
> 
> produces
> 
> ```
> {'oint': memmap([0., 0., 0., ..., 0., 0., 0.])}
> ```
> 
> Somewhere, the 'p' is deleted. Please check that this PR fixes it.

I thought this was fixed! :frowning: Will take a look at it

jhaux committed:
:bug: Further cleanup

jhaux committed:
Merge branch 'master' into port_evaldatafolder

jhaux committed:
:bug: Stops TemplateEvalHook from importing callbacks

jhaux committed:
Merge branch 'master' into port_evaldatafolder

pesser committed:
eval_callbacks must be dict

remove useless conversion from string to dict

pesser committed:
Update pipeline.py

after talk with jhaux, agreed to log only exceptional case

pesser commented on edflow/eval/pipeline.py:
```diff
@@ -1194,7 +883,7 @@ def main():
     callbacks = cbargs2cbdict(args.callback)
 
-    standalone_eval_csv_file(args.csv, callbacks, additional_kwargs, args.other_config)
+    standalone_eval_meta_dset(args.csv, callbacks, additional_kwargs, args.other_config)
```
update argparser, documentation and tests to reflect the change from csv to meta yaml and we're ready to merge

pesser reviewed (changes requested):
this makes everything much cleaner. one minor change and we can merge

jhaux committed:
:memo: Fixes edeval help text and changes argument names

jhaux committed:
Merge branch 'port_evaldatafolder' of github.com:jhaux/edflow into port_evaldatafolder

pesser reviewed (approved):
:sunglasses:
